File: 2.1 Algorithmic Toolbox/Week 3 Greedy Algorithms/02_greedy_algorithms_starter_files/fractional_knapsack/fractional_knapsack.py
# Uses python3
import sys

def get_optimal_value(capacity, weights, values):
    value = 0.
    # write your code here
    vpw = [x/y for x, y in zip(values, weights)]
    vpw, weights, values = zip(*sorted(zip(vpw, weights, values), reverse=True))
    print(vpw)
    print(values)
    print(weights)
    
    rem = capacity
    for i in range(len(vpw)):
        item = rem
        if item >= weights[i]:
            item = weights[i]
            value += item * vpw[i]
            rem -= item
        else:
            value += item * vpw[i]
            rem -= item

    return value


if __name__ == "__main__":
    data = list(map(int, sys.stdin.read().split()))
    n, capacity = data[0:2]
    values = data[2:(2 * n + 2):2]
    weights = data[3:(2 * n + 2):2]
    opt_value = get_optimal_value(capacity, weights, values)
    print("{:.10f}".format(opt_value))

File: 2.1 Algorithmic Toolbox/Week 3 Greedy Algorithms/02_greedy_algorithms_starter_files/fractional_knapsack/fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

using std::vector;

vector<int> get_index(vector<double>& ratio){
	vector<int> idx(ratio.size());
	std::iota(idx.begin(),idx.end(),0);
	std::sort(idx.begin(), idx.end(), [&](double a, double b){return ratio[a] > ratio[b];});
	return idx;
}

double get_optimal_value(int capacity, vector<int> weights, vector<int> values) {
  double value = 0.0;

  vector<double> vpw(weights.size());
  for(int i = 0; i < weights.size(); i++){
	  vpw[i] = double(values[i]) / double(weights[i]);
  }

  int rem = capacity;
  vector<int> index = get_index(vpw);
  for(int i = 0; i < index.size(); i++){
	  int item = rem;
	  if(item >= weights[index[i]]){
		  item = weights[index[i]];
		  value += item * vpw[index[i]];
		  rem -= item;
	  }
	  else{
		  value += item * vpw[index[i]];
		  rem -= item;
	  }
  }

  return value;
}

int main() {
  int n;
  int capacity;
  std::cin >> n >> capacity;
  vector<int> values(n);
  vector<int> weights(n);
  for (int i = 0; i < n; i++) {
    std::cin >> values[i] >> weights[i];
  }

  double optimal_value = get_optimal_value(capacity, weights, values);

  std::cout.precision(10);
  std::cout << optimal_value << std::endl;

  system("pause");
  return 0;
}
